libsignal-service/src/provisioning.rs

```diff
@@ -0,0 +1,328 @@
+    pub fn new(ctx: Context) -> Result<Self, Error> {
+        let key_pair = libsignal_protocol::generate_key_pair(&ctx)?;
+        Ok(Self { ctx, key_pair })
+    }
+
+    pub fn from_key_pair(ctx: Context, key_pair: KeyPair) -> Self {
+        Self { ctx, key_pair }
+    }
+
+    pub fn public_key(&self) -> PublicKey {
+        self.key_pair.public()
+    }
+
+    pub fn decrypt(
+        &self,
+        provision_envelope: ProvisionEnvelope,
+    ) -> Result<ProvisionMessage, ProvisioningError> {
+        let master_ephemeral = PublicKey::decode_point(
+            &self.ctx,
+            &provision_envelope.public_key.expect("no public key"),
+        )?;
+        let body = provision_envelope
+            .body
+            .expect("no body in ProvisionMessage");
+        if body[0] != 1 {
+            return Err(ProvisioningError::InvalidData {
+                reason: "Bad version number".into(),
+            });
+        }
+
+        let iv = &body[IV_OFFSET..(IV_LENGTH + IV_OFFSET)];
+        let mac = &body[(body.len() - 32)..];
+        let cipher_text = &body[16 + 1..(body.len() - CIPHER_KEY_SIZE)];
+        let iv_and_cipher_text = &body[0..(body.len() - CIPHER_KEY_SIZE)];
+        debug_assert_eq!(iv.len(), IV_LENGTH);
+        debug_assert_eq!(mac.len(), 32);
+
+        let agreement =
+            master_ephemeral.calculate_agreement(self.key_pair.private())?;
+        let hkdf = libsignal_protocol::create_hkdf(&self.ctx, 3)?;
+
+        let shared_secrets = hkdf.derive_secrets(
+            64,
+            &agreement,
+            &vec![],
+            b"TextSecure Provisioning Message",
+        )?;
+
+        let parts1 = &shared_secrets[0..32];
+        let parts2 = &shared_secrets[32..];
+
+        use hmac::{Hmac, Mac, NewMac};
+        use sha2::Sha256;
+        let mut verifier = Hmac::<Sha256>::new_varkey(&parts2)
+            .expect("Hmac can take any size key");
+        verifier.update(&iv_and_cipher_text);
+        let our_mac = verifier.finalize().into_bytes();
+        debug_assert_eq!(our_mac.len(), mac.len());
+        if &our_mac[..32] != mac {
+            return Err(ProvisioningError::InvalidData {
+                reason: "wrong MAC".into(),
+            });
+        }
+
+        use aes::Aes256;
+        // libsignal-service-java uses Pkcs5,
+        // but that should not matter.
+        // https://crypto.stackexchange.com/questions/9043/what-is-the-difference-between-pkcs5-padding-and-pkcs7-padding
+        use block_modes::{block_padding::Pkcs7, BlockMode, Cbc};
+        let cipher = Cbc::<Aes256, Pkcs7>::new_var(&parts1, &iv)
+            .expect("initalization of CBC/AES/PKCS7");
+        let input = cipher.decrypt_vec(cipher_text).expect("decryption");
+
+        Ok(prost::Message::decode(Bytes::from(input))?)
+    }
+}
+
+#[pin_project]
+pub struct ProvisioningPipe<WS: WebSocketService> {
+    ws: WS,
+    #[pin]
+    stream: WS::Stream,
+    provisioning_cipher: ProvisioningCipher,
+}
```

Comment:
Maybe this should be a state machine with types for every step, to make it very Rusty.

Reply:
It can pass the stream, cipher and ws through every step, but the type encodes the current step in the process.

Reply:
Does using a state machine give us anything in terms of correctness (by design the public API can't be used incorrectly) or make the code simpler? If not, I wouldn't bother because it'll add more work and increase cognitive load (instead of needing to understand a single type with a couple methods, you need to mentally parse a state machine and its various transitions).